keras_nlp/tokenizers/word_piece_tokenizer_trainer.py

```diff
@@ -0,0 +1,124 @@
+    This function can be used to train a Word Piece vocabulary from an input
+    dataset, or a list of filenames.
+
+    Args:
+        data: A tf.data.Dataset, or a list of filenames.
+        vocabulary_size: The maximum size of a vocabulary to be trained.
+        vocabulary_output_file: The location to write a vocabulary file.
+        lowercase: If true, the input text will be first lowered before
+            tokenization.
+        strip_accents: If true, all accent marks will be removed from text
+            before tokenization.
+        split: If true, input will be split on whitespace and punctuation
+            marks, and all punctuation marks will be kept as tokens. If false,
+            input should be split ("pre-tokenized") before calling the
+            tokenizer, and passed as a dense or ragged tensor of whole words.
+            `split` is required to be`True` when `data` is a list of filenames.
+        suffix_indicator: The characters prepended to a wordpiece to indicate
+            that it is a suffix to another subword.
+        reserved_tokens: A list of tokens that must be included in the vocabulary.
+
+    Returns:
+        Returns a list of vocabulary terms.
+    """
+    # Read data files.
+
+    if not isinstance(data, (list, tf.data.Dataset)):
+        raise ValueError(
+            "The `data` argument must be either `tf.data.Dataset` or `list`. "
+            f"Recieved: {type(data)}."
+        )
+    if isinstance(data, list):
+        if not split:
+            raise ValueError(
+                "When learning a vocab from files, `split` must be `True`. "
+                "To compute a vocabulary with custom split rules, load your "
+                "data as a dataset, split it, and pass it to "
+                "`compute_word_piece_vocabulary()` with split=False."
+            )
+        data = tf.data.TextLineDataset(data)
+
+    def preprocess(text):
+        """Takes in a dataset element and preprocesses it."""
+        # Check for correct types.
+        if text.dtype != tf.string:
+            raise ValueError(
+                "The dataset elements in `data` must have string dtype. "
+                f"Recieved: {text.dtype}."
+            )
+        # Preprocess, lowercase, strip and split input data.
+        if text.shape.rank == 0:
+            text = tf.expand_dims(text, 0)
+        if lowercase:
+            text = tf_text.case_fold_utf8(text)
+        if strip_accents:
+            # Normalize unicode to NFD, which splits out accent mark characters.
+            text = tf_text.normalize_utf8(text, "NFD")
+            # Remove the accent marks.
+            text = tf.strings.regex_replace(text, r"\p{Mn}", "")
+        if split:
+            text = tf_text.regex_split(
+                text,
+                delim_regex_pattern=WHITESPACE_AND_PUNCTUATION_REGEX,
+                keep_delim_regex_pattern=PUNCTUATION_REGEX,
+            )
+        return text
+
+    words_data = data.map(preprocess)
```

Comment:
make sure to add `num_parallel_calls=tf.data.AUTOTUNE` as an argument (this is a persistent gotcha with tf.data). That will allow multi threaded pre tokenization.